Update StatsService to be a singleton that polls the api at a regular…

aljones15 opened the pull request:
Update StatsService to be a singleton that polls the api at a regular intervals.

- Constructor accepts a poll time.
- Service is a singleton that checks the window for a previous instance.
- Update updates subscribers after getting latest data.
- Add a set called subscribers to prevent duplicate subscriptions.
- Add a lastTime get function.

@mattcollier this is a clean up of this branch it contains the new StatService and also updated test project with out the vue files in the history that are now in bedrock-vue-stats.

ChartController was added to this PR.

This can be a release.

aljones15 committed:
Update StatsService to be a singleton that polls the api at a regular interval.

- Constructor accepts a poll time.
- Service is a singleton that checks the window for a previous instance.
- Update updates subscribers after getting latest data.
- Add a set called subscribers to prevent duplicate subscriptions.
- Add a lastTime get function.

Update index.js removing unneeded chart.js reference.

Remove cherrypick from files.

Cherrypick Home.vue.

Remove cherrypick from home.

Add home vue.

Upgrade StatsService.

aljones15 committed:
Add bedrock-vue-stats to project.

aljones15 committed:
Rebase out gmail address.

aljones15 committed:
Update charts to use realtime bool.

aljones15 committed:
Add a ChartController.

Start a formatter for data.

Abstract pie charts into ChartController and add loading to StatService.

Switch pie charts to ChartController.

aljones15 committed:
Implement ChartController for pie and realtime charts.

aljones15 committed:
Update jsdoc in ChartController.

aljones15 committed:
Remove gb and formatter from StatsService.

aljones15 committed:
Round up on max disk space.

mattcollier reviewed (changes requested):
eslint-config-digitalbazaar@1.8 includes the jsdocs rules, please enable that and lint all these files.

Please dedicate at least 30 mins to doing a self-review on this diff.

ChartController.js is missing copyright header.

aljones15 committed:
Add full docs.

aljones15 committed:
Fix format of links in read me.

Space links in README.

aljones15 committed:
Add a script to generarte docs.

aljones15 committed:
Add copyright to docs.

aljones15 committed:
Fix a few typos expand docs.

aljones15 committed:
Update ChartController copyright date to just 2019.

aljones15 committed:
Implement a better way of dterming how many cores are being used in CPU Usage.

aljones15 committed:
Change StatsService to send all data on update.

aljones15 committed:
Refactor ChartController so charts can see all data from update.

gannan08 commented:
Unable to test completely. Many errors.
![Screenshot_2019-03-14_14-23-48](https://user-images.githubusercontent.com/13918179/54381743-e2f6a680-4664-11e9-950d-433bab35754c.png)

aljones15 commented:
> Unable to test completely. Many errors.
> ![Screenshot_2019-03-14_14-23-48](https://user-images.githubusercontent.com/13918179/54381743-e2f6a680-4664-11e9-950d-433bab35754c.png)

all of those come from chartjs plugin streaming... looking into but on both mac and linux it works for me. going to try and delete node_modules. what browser did you use?

gannan08 commented:
> all of those come from chartjs plugin streaming... looking into but on both mac and linux it works for me. going to try and delete node_modules. what browser did you use?


Google Chrome
Version 72.0.3626.119 (Official Build) (64-bit)

aljones15 commented:
@gannan08  ok man one more time issue was with bedrock-vue-stats and chart.js. chart.js updated to 2.8 this morning breaking functionality with the plugin. let me know it should work now.

p.s. I did make one more commit tonight mostly because I figured out a way of further cutting down on the complexity of the subscription and update process.

aljones15 committed:
Reduce the number of loops needed to created chart data.

Change test data to update every 3 seconds.

gannan08 commented on .eslintrc.js:
```diff
@@ -3,6 +3,6 @@ module.exports = {
   },
   plugins: ['jsdoc'],
-  extends: 'eslint-config-digitalbazaar',
+  extends: 'eslint-config-digitalbazaar/jsdoc',
```
This eslint-config name breaks the pattern I conceived in my mind. It seems the pattern was `eslint-config-digitalbazaar/<env>`, eg. **vue:** `eslint-config-digitalbazaar/vue`. This does not block this PR.

mattcollier replied:
I think the pattern is that the thing at the end is the plugin, not the environment.

gannan08 replied:
So if we wanted to use jsdocs in our vue environment then it would be...

```
extends: ['eslint-config-digitalbazaar/vue','eslint-config-digitalbazaar/jsdoc']
```

Is the above statement correct?

mattcollier replied:
I'm very pleased that you have brought up this topic.  What you are suggesting is the way I think it should work, except, maybe you would actually go one step further:
```
extends: ['eslint-config-digitalbazaar', 'eslint-config-digitalbazaar/vue','eslint-config-digitalbazaar/jsdoc']
```
keep in mind here that these *could* be shortened to just digitalbazaar, digitalbazaar/vue.  Somewhere along the line it was decided that we would use the verbose full names.

At any rate, the way the `vue` and `jsdoc` rules are currently setup do not accurately reflect what I have written above.  `vue` and `jsdoc` already extend the base linter eslint-config-digitalbazaar as shown here (the base rules are represented here by index.js): https://github.com/digitalbazaar/eslint-config-digitalbazaar/blob/master/vue.js#L5

So, I don't know if the current rule setup causes any problems or confusion on the part of the linter.  Also, since eslint rules are additive,  it's common that we would have eslint-config-digitalbazaar at the root of the project, then eslint-config-digitalbazaar/vue in the components folder. 

So, thus far, just using 'eslint-config-digitalbazaar-vue' has just been working the way people expect, without having the be explicit about adding the base 'eslint-config-digitalbazaar' rules.

aljones15 replied:
why don't we move this discussion to eslint-config-digitalbazaar maybe we could make an issue separate rules so we can extend by composition i.e.

extends: ['digitalbazzar', 'digitalbazaar/vue'] etc.

something tells me this Pr is about to get pretty big.

gannan08 commented on ChartController.js:
```diff
@@ -0,0 +1,148 @@
+ * @property {number} gb - the number of bytes in a gigabyte.
+ */
+const ChartUnit = {
+  gb: Math.pow(2, 30)
```
Gigabytes and any other unit specific information should not be determined here. The unit declarations should be passed in to this module. What if one wanted to use liters or gallons? This structure essentially hardcodes that the units are some denomination of bytes.

aljones15 replied:
so why don't I refactor and move units out of the ChartController and then a Component could import a ByteUnits object with the conversions.

gannan08 commented on ChartController.js:
```diff
@@ -0,0 +1,148 @@
+};
+
+/**
+ * Receives all the data from each update and get it's ready for the other
```
Typo, "its" not "it's".

aljones15 replied:
Actually I think it should just be it. will correct this in a moment.

gannan08 commented on StatsService.js:
```diff
@@ -7,12 +7,118 @@ import axios from 'axios';
+   * @returns {StatsService} Either an existing one or a new one.
+   */
+  constructor({poll = 1000} = {}) {
+    if(!window.bedrock) {
```
Why are we using globals here? Is this the only path forward that you found for completing this section of code?

aljones15 replied:
what would you suggest here? How for instance did angular handle singletons? there has to be a place to look for the singleton so I used window. more than happy to revise.

gannan08 commented on ChartController.js:
```diff
@@ -0,0 +1,148 @@
+  }
+  /**
+   * Called by StatsService on each update.
+   * This function is async because we do not want it blocking.
```
Don't think this comment is necessary. Javascript is single threaded, therefore you almost never want to block the main loop.

aljones15 replied:
I actually think this comment is necessary because by being async is goes on the async que which means that the StatsService is able to issue all updates to a que instead of awaiting them.

gannan08 commented:
Additionally, the conversion from seconds to minutes is not clean. I'd expect the entire chart to adjust accordingly based off of units. All it does is change the labels from every 5 seconds to every minute. I expected the chart to update every minute or a few seconds. The chart should condense in minute view and then expand in seconds view, if this is confusing then ping me on a chat line.

mattcollier commented:
> Additionally, the conversion from seconds to minutes is not clean. I'd expect the entire chart to adjust accordingly based off of units. All it does is change the labels from every 5 seconds to every minute. I expected the chart to update every minute or a few seconds. The chart should condense in minute view and then expand in seconds view, if this is confusing then ping me on a chat line.

@aljones15 @gannan08 I do recommend a voice conversation be had about this.  Keeping in mind that we want to have the pan/zoom capability in place after a 1.0 release.  It is my hope that the pan/zoom plugin is going to handle adding ticks etc for us automagically.  It is my feeling that we don't currently have any use case for this timescale (min, second) drop down box.

gannan08 commented:
> @aljones15 @gannan08 I do recommend a voice conversation be had about this. Keeping in mind that we want to have the pan/zoom capability in place after a 1.0 release. It is my hope that the pan/zoom plugin is going to handle adding ticks etc for us automagically. It is my feeling that we don't currently have any use case for this timescale (min, second) drop down box.

@mattcollier @aljones15 I agree that a voice conversation will be needed to further explain my thoughts. If we're not using the drop down box, then I'm assuming we'd have some predetermined interval.

aljones15 committed:
Remove units from ChartController.

aljones15 committed:
Remove StatsService from window & change subscriptions to promises.

- Update docs in ChartController.
- Add @typedef for Chart.
- Store StatsService singleton inside module.
- Subscriptions are now Promises.
- Update docs in StatsService.
- Correct use .bind to properly store new function.

aljones15 committed:
Replace updater in ChartController with subscription for better clarity.

gannan08 commented on test/components/Home.vue:
```diff
@@ -23,34 +59,99 @@
+    }
+  },
+  mounted() {
+    this.$set(this, 'cpu', new ChartController(
```
Is this the intended interaction with the API?

- Why are we using `$set()` on variables that are created in `data()`?
- For `format` it seems to me that `prefix` is quite misleading. I believe
  this function provides the location of the data in an object for use of
  display if I am not mistaken. How can we name this better to fit its
  true purpose?

gannan08 replied:
Also, don't know how Vue reactivity works with the instantiation of a class. Do the proper getters and setters work?

aljones15 replied:
@gannan08 $set guarantees an update. Vue's reactivity actually will not update if you mutate the data and while I believe you are correct that replacing null with a class will be noticed and cause an update I would prefer that we adopt using this.$set in all cases (I have worked on apps where people pushed to an array which resulted in vue not updating etc.) using $set is just good form IMHO. Every Vue project I have worked on I have observed cases where people do not understand Vue's reactivity so I use $set in all cases in order to be sure. Other advantages of $set:
- no need for Object spread polyfils
- no need for array spread polyfils
- less use of Object.assign etc.

I agree on prefix is a bad name maybe one of these options:
- getter
- prepare or preparer
- formatter

I think prepare is good actually.

aljones15 replied:
as for the getters and setters I guess you mean the class? does seem to be working not having any issues. Good concern though.

gannan08 commented on StatsService.js:
```diff
@@ -7,12 +7,119 @@ import axios from 'axios';
+// instance stores the singleton
+let instance = null;
 
+class StatsService {
```
We are going to change the singleton pattern here. The reason behind the change is because we want to allow for one to create instantiate new objects if they want. So, the change will require setting up a singleton in the `index.js` file and exporting that.

Here's an example of the pattern we want to achieve:
https://github.com/digitalbazaar/bedrock-web-store/blob/master/index.js

aljones15 replied:
that is beautiful and will do that first thing this morning.

aljones15 committed:
Export singleton in index.js & Rename prefix to prepare.

- Remove all Singleton logic from StatsService module.
- Add start and stop methods to StatsService.
- Change Prefix to Prepare
- Change ChartController to take StatsService as a paramater.
- Update docs.
